### DISCUSSION TITLE
K shortest paths with negative weights, all_shortest_paths to be used with bellman ford

### DISCUSSION THREAD

**Discussion by maratos92** [networkx/networkx#7458] [Algorithms]
Hi, 

I have a MultiDiGraph in which I need to find k paths from one node to the leaves (or a custom list of nodes) in order of their weight. Graph has edges with negative weights. The only possible solution in my case would be to find all simple paths and then select the shortest out of them manually because   all_shortest_paths works with dijkstra algorithm only. In this case the algorithm becomes very greedy and requires manual manipulations. 
I get this error when trying to find all_shortest_paths: 
Error: Contradictory paths found: negative weights?

**Comment by dschult** [comment]
Once a cycle is found that has negative weight, the distance between two nodes is `-inf` because you could keep traveling around and around the same negative loop. So you will not find a shortest_path and thus djikstra raises an exception.

One not-quite-work-around for the negative edge weight problem is obtained by forming a new edge weight attribute that is a constant M plus the original weight. Make M big enough that all edge weights are positive. Then the concept of shortest_path will be defined and you can continue.  Note that the length of the paths now depends on the number of hops and the sum of the original weights. So this may not be what you want in the end. 

Without manipulating edge weights, it seems possible to iterate over all simple paths and only store the k shortest. For large graphs this might be slow, but it shouldn't crash from building too many paths to hold in memory and it might work in your case.

Finally, you should think about what you mean by the "k path in order of their weight".  What is it that you really need?  The negative weights in your case apparently allow infinitely negative weight paths. Think hard about what your weights (and the path weights) will mean to you and maybe there will be another problem close-by that gives you what you actually want instead of infinities.

**Reply by maratos92** [reply]
Thanks for your comment! I also thought of adding a constant to all edges and even tried it but then figured out it is not going to work for my case. My idea is to write code to identify k most critical paths in a project schedule. Construction schedules can be quite detailed, I've seen integrated construction schedules with more than 20k nodes. Aim is to select a target activity (for example project completion) and find activities contributing to it's delay, so people working on a project can focus on finishing particular works first and other works can have less priority. There's planning software like Primavera P6 or Asta Power Projects, but functionality and outputs aren't ideal, you can select only one target activity at a time or software identifies constrained paths as critical etc. Weights would be the "free slack or float" between activities and sometimes activities planned one after the other are out of sequence for example one activity is linked Finish-to-Start after another but it is actually completed earlier than the predecessor even starts. In the end I still need to calculate the whole path cost and in this case I may have negative weights. I ended up fixing these weights and replacing them by software computed "free float" which is never negative, it is not ideal but my dijkstra heuristic shows better results than existing software in the end. 
I also tried all simple paths, but average schedule is around 1k-5k nodes and I might have many roots and leaves to check between each other + sometimes these graphs are quite dense so it takes time, at least I did few tests with 3k nodes and it was computing longer than half an hour, in my case it's a waste of time and memory. Maybe with cugraph it would work better but all simple paths aren't available for that class yet, not to mention that you need to have machine with GPU, set up CUDA and so on. In any case thanks for explanation, I've gone through Yen's algorithm and even found few academic papers with improvements but it looks like this problem still can be only solved by identifying all paths in a graph
